crates/core/src/driver/mod.rs

```diff
@@ -117,238 +136,236 @@ where
         }
     }
 
-    pub fn execute_input(
+    pub fn handle_input(
         &mut self,
+        metadata: &Metadata,
+        case_idx: CaseIdx,
         input: &Input,
         node: &T::Blockchain,
     ) -> anyhow::Result<(TransactionReceipt, GethTrace, DiffMode)> {
-        tracing::trace!("Calling execute_input for input: {input:?}");
+        let deployment_receipts =
+            self.handle_contract_deployment(metadata, case_idx, input, node)?;
+        self.handle_input_execution(case_idx, input, deployment_receipts, node)
+    }
 
-        let nonce = node.fetch_add_nonce(input.caller)?;
+    /// Handles the contract deployment for a given input performing it if it needs to be performed.
+    fn handle_contract_deployment(
+        &mut self,
+        metadata: &Metadata,
+        case_idx: CaseIdx,
+        input: &Input,
+        node: &T::Blockchain,
+    ) -> anyhow::Result<HashMap<ContractInstance, TransactionReceipt>> {
+        let span = tracing::debug_span!(
+            "Handling contract deployment",
+            ?case_idx,
+            instance = ?input.instance
+        );
+        let _guard = span.enter();
+
+        let mut instances_we_must_deploy = IndexMap::<ContractInstance, bool>::new();
```

Comment:
I guess the reason for the indexmap is that iteration needs to be predictable and exactly reproducible?